src/scripts/main.js

```diff
@@ -1,7 +1,212 @@
+const RARE_CELL_CHANCE = 10;
 
-// Write your code here
+const inputHandler = ({ key }) => {
+  moveCells(key);
+};
+
+refs.startBtn.addEventListener('click', newGame);
+
+function newGame() {
+  resetField();
+  addCell(getRandomCell(), getToken());
+  addCell(getRandomCell(), getToken());
+  window.addEventListener('keydown', inputHandler);
+}
+
+function getRandomCell() {
+  if (!hasFreeSpace()) {
+    return -1;
+  }
+
+  let randomCell = getRandomNum(FIELD_LENGTH);
+
+  while (refs.fieldCells[randomCell].textContent !== '') {
+    randomCell = getRandomNum(FIELD_LENGTH);
+  }
+
+  return randomCell;
+}
+
+function getToken() {
+  return getRandomNum(100) <= RARE_CELL_CHANCE
+    ? RARE_NEW_CELL_VALUE
+    : NEW_CELL_VALUE;
+}
+
+function getRandomNum(maxVal) {
+  return Math.floor(Math.random() * maxVal);
+}
+
+function hasFreeSpace() {
+  return [...refs.fieldCells].some((cell) => cell.textContent === '');
+}
+
+function resetField() {
+  const { messageStart, messageLose, messageWin, score, fieldCells } = refs;
+
+  score.textContent = '0';
+
+  [messageStart, messageLose, messageWin].forEach(({ classList }) =>
+    classList.add('hidden'),
+);
+
+  fieldCells.forEach((cell) => {
+    cell.classList.remove(cell.classList[1]);
+    cell.textContent = '';
+  });
+}
+
+function updateStartButton() {
+  const { startBtn } = refs;
+
+  startBtn.classList.remove('start');
+  startBtn.classList.add('restart');
+  startBtn.textContent = 'Restart';
+}
+
+function moveCells(keyPressed) {
+  updateStartButton();
+
+  !rearrangeCells(keyPressed) && addCell(getRandomCell(), getToken());
+
+  !hasFreeSpace() && !hasPossibleMoves() && finishGame(refs.messageLose);
+}
+
+function addCell(cellPosition, cellValue) {
+  if (cellPosition === -1) {
+    return;
+  }
+
+  const { fieldCells } = refs;
+
+  fieldCells[cellPosition].classList.add(`field-cell--${cellValue}`);
+  fieldCells[cellPosition].textContent = cellValue;
+}
+
+function finishGame(endGameMessage) {
+  endGameMessage.classList.remove('hidden');
+  window.removeEventListener('keydown', inputHandler);
+}
+
+function updateScore(points) {
+  const { score } = refs;
+  const currentPoints = +score.textContent;
+
+  score.textContent = currentPoints + points;
+}
+
+function rearrangeCells(direction) {
+  const { fieldCells, fieldRows } = refs;
+
+  const rows =
+    direction === 'ArrowLeft' || direction === 'ArrowRight'
+      ? [...fieldRows]
+      : createFieldColumns(fieldCells);
+
+  const initialRowValues = rows
+    .map((row) => [...row.children].map((cel) => cel.textContent))
+    .flat();
+
+  rows.forEach((row) => {
+    const initialCells =
+      direction === 'ArrowLeft' || direction === 'ArrowUp'
+        ? [...row.children]
+        : [...row.children].reverse();
+
+    const filledCellsValues = initialCells
+      .filter((cell) => cell.textContent !== '')
+      .map(({ textContent }) => textContent);
+
+    filledCellsValues.forEach((cell, i) => {
```

Comment:
This merging logic does not remove the class associated with the old value from the cell. Ensure that the cell's classes are updated correctly when merging.